### DISCUSSION TITLE
How whisper managed previous context during training?

### DISCUSSION THREAD

**Discussion by joseluis-recog** [openai/whisper#2476] [Q&A]
I was wondering whether Whisper's developers make use of a padding token during batch training. Specifically, I’ve been experimenting with feeding the context (the tokens situated between <|startofprev|> and <|startoftranscript|>) during training. To facilitate batching, I need to pad these contexts to match the length of the longest one in the batch.

However, I couldn’t find any documentation or references regarding the use of padding tokens during Whisper’s training process. I’ve tried various padding approaches, such as padding on the left, padding on the right, using -100, and even using token ID 50256 (which corresponds to the empty string, ""). In all cases, Whisper seems to output random, nonsensical tokens in response to the padded inputs.

This behavior leads me to suspect that instead of padding, Whisper's developers might have truncated the contexts to the shortest length in the batch during training. If that’s the case, it would explain why the model doesn’t recognize any specific padding pattern.

If anyone has insights into how this was handled during training or knows the correct approach, it would be incredibly helpful!

**Comment by Advait251206** [comment]
Whisper's training details are not fully public, so there is no official description of exactly how previous-context sequences were batched. However, based on the architecture, the training code released by OpenAI, and the behavior you've observed, a few things can be inferred.

## Short Answer

Whisper was almost certainly **not trained to interpret arbitrary padding tokens inside the `<|startofprev|> ... <|startoftranscript|>` context region**.

If you insert padded tokens into the prompt context, the model may treat them as real tokens and generate unpredictable outputs.

---

## How Whisper Uses Previous Context

At inference time, Whisper can receive a prompt like:

```text
<|startofprev|>
previous transcript tokens...
<|startoftranscript|>
<|en|>
<|transcribe|>
<|notimestamps|>
```

The decoder simply sees one long token sequence.

There is no special masking mechanism that says:

```text
"Ignore everything before here because it's padding"
```

as you would have in models trained with explicit attention masks.

---

## Why Padding Causes Problems

Suppose your batch contains contexts of different lengths:

```text
Sample A: 100 tokens
Sample B: 20 tokens
```

and you pad Sample B to length 100.

If the padding token is:

```python
50256
```

(or any other token)

the decoder still receives those tokens as valid input unless they are properly masked during loss computation and attention.

From the model's perspective:

```text
<|startofprev|> real text ... PAD PAD PAD PAD ...
```

is simply another sequence it has likely never seen during training.

This often results in:

- hallucinations
- unstable loss
- random token generation
- degraded conditioning

which matches your observations.

---

## Did OpenAI Use Padding?

Almost certainly yes at the batch level, because transformer training requires batching.

The more important question is:

### Was padding attended to?

Most likely not.

Typical sequence-to-sequence training would use:

```python
attention_mask
```

and

```python
labels = -100
```

for padded target positions.

The model learns to ignore padded regions through masking rather than through the semantic meaning of a padding token.

---

## What Hugging Face Does

The Hugging Face Whisper implementation typically uses:

```python
labels[labels == pad_token_id] = -100
```

for the target sequence.

For decoder inputs:

```python
attention_mask
```

controls which positions participate in attention.

The padding token itself is not expected to carry meaning.

---

## Why `<|startofprev|>` Is Tricky

The previous-context region is different from ordinary decoder padding.

For example:

```text
<|startofprev|>
token1 token2 token3 PAD PAD PAD
<|startoftranscript|>
...
```

The decoder may still attend to those PAD tokens because they occur inside what it believes is meaningful prompt text.

Unless you explicitly mask them out, they become part of the prompt.

---

## Recommended Approaches

### Option 1: Dynamic Batching (Best)

Group examples with similar prompt lengths.

```python
batch = [
    context_len ≈ 50
]
```

This minimizes padding.

---

### Option 2: Left-Truncate Context

Instead of padding long histories:

```python
context = context[-N:]
```

Keep only the most recent tokens.

This is generally how Whisper prompting is used in practice.

---

### Option 3: Attention Masking

If using the Hugging Face implementation, ensure:

```python
decoder_attention_mask
```

properly masks padded prompt tokens.

Without masking, padding becomes part of the prompt.

---

### Option 4: No Previous Context During Fine-Tuning

Many successful Whisper fine-tuning pipelines simply train on:

```text
<|startoftranscript|>
<|language|>
<|transcribe|>
...
```

without supplying `<|startofprev|>` at all.

Unless your application specifically relies on conversational or long-form context, this is often the simplest and most stable setup.

---

## About Using `-100`

`-100` should only be used in:

```python
labels
```

for loss masking.

It should **not** be inserted into decoder input IDs.

For example:

```python
labels[pad_positions] = -100
```

✓ Correct

```python
input_ids[pad_positions] = -100
```

✗ Incorrect

---

## My Guess Based on Your Results

The fact that:

- left padding fails,
- right padding fails,
- `50256` fails,
- and outputs become nonsensical,

suggests the model is attending to those padded prompt tokens as if they were real context.

This is consistent with Whisper not being trained to interpret padding inside the `<|startofprev|>` section.

If you're fine-tuning with previous-context prompts, the safest strategy is usually:

1. Use attention masks for padding.
2. Bucket sequences by context length.
3. Truncate long contexts instead of heavily padding short ones.
4. Avoid relying on any special "padding token semantics" in the prompt itself.

That's much closer to how Whisper was likely trained and used internally.
